src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnectionImpl.java

```diff
@@ -1412,16 +1425,24 @@ private Object doPrivilegedOperation(final int operation,
         try {
             PrivilegedOperation op = new PrivilegedOperation(operation, params);
             if (acc == null) {
-                try {
+                // No ACC, therefore no SM. May have a Subject:
+                if (subject != null) {
+                    return Subject.doAs(subject, op);
+                } else {
                     return op.run();
-                } catch (Exception e) {
-                    if (e instanceof RuntimeException)
-                        throw (RuntimeException) e;
-                    throw new PrivilegedActionException(e);
                 }
             } else {
+                // ACC is present, we have a Subject and SM is permitted:
                 return AccessController.doPrivileged(op, acc);
             }
+        } catch (Exception e) {
+            if (e instanceof SecurityException) {
+                throw (SecurityException) e;
+            } else if (e instanceof RuntimeException) {
+                throw (RuntimeException) e;
+            } else {
+                throw new PrivilegedActionException(e);
+            }
```

Comment:
I assume there no chance that `Exception e` may already be a `PrivilegedActionException` here. You coud avoid casts by using instanceof patterns.

```suggestion
            if (e instanceof SecurityException se) {
                throw se;
            } else if (e instanceof RuntimeException rte) {
                throw  rte;
            } else {
                throw new PrivilegedActionException(e);
            }
```
Alternatively maybe a switch expression with a default case could be used?

Reply:
ok yes will do that.  SecurityException was another carryover from before simplifying the updated calls, think it's no longer needed.